pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -0,0 +1,512 @@
+            (, underlyingAmounts[i], wrappedAmounts[i]) = _vault.erc4626BufferWrapOrUnwrap(
+                BufferWrapOrUnwrapParams({
+                    kind: kind,
+                    direction: WrappingDirection.WRAP,
+                    wrappedToken: wrappedToken,
+                    amountGivenRaw: amountsIn[i],
+                    limitRaw: limits[i],
+                    userData: params.userData
+                })
+            );
+
+            if (isStaticCall == false && kind == SwapKind.EXACT_OUT) {
+                // If SwapKind of wrap is EXACT_OUT, the limit of underlying tokens was taken from the user, so the
+                // difference between limit and exact underlying amount needs to be returned to the sender.
+                _vault.sendTo(underlyingToken, params.sender, limits[i] - underlyingAmounts[i]);
+            }
+        }
+    }
+
+    /***************************************************************************
+                                   Nested pools
+    ***************************************************************************/
+
+    function removeLiquidityProportionalFromNestedPools(
+        address parentPool,
+        uint256 exactBptAmountIn,
+        uint256[] memory minAmountsOut,
+        bytes memory userData
+    ) external returns (address[] memory tokensOut, uint256[] memory amountsOut) {
+        (tokensOut, amountsOut) = abi.decode(
+            _vault.unlock(
+                abi.encodeWithSelector(
+                    CompositeLiquidityRouter.removeLiquidityProportionalFromNestedPoolsHook.selector,
+                    RemoveLiquidityHookParams({
+                        sender: msg.sender,
+                        pool: parentPool,
+                        minAmountsOut: minAmountsOut,
+                        maxBptAmountIn: exactBptAmountIn,
+                        kind: RemoveLiquidityKind.PROPORTIONAL,
+                        wethIsEth: false,
+                        userData: userData
+                    })
+                )
+            ),
+            (address[], uint256[])
+        );
+    }
+
+    function removeLiquidityProportionalFromNestedPoolsHook(
+        RemoveLiquidityHookParams calldata params
+    ) external nonReentrant onlyVault returns (address[] memory tokensOut, uint256[] memory amountsOut) {
+        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(params.pool);
+
+        (, uint256[] memory parentPoolAmountsOut, ) = _vault.removeLiquidity(
+            RemoveLiquidityParams({
+                pool: params.pool,
+                from: params.sender,
+                maxBptAmountIn: params.maxBptAmountIn,
+                minAmountsOut: new uint256[](parentPoolTokens.length),
+                kind: params.kind,
+                userData: params.userData
+            })
+        );
+
+        for (uint256 i = 0; i < parentPoolTokens.length; i++) {
+            address childPool = address(parentPoolTokens[i]);
```

Comment:
```suggestion
            address childToken = address(parentPoolTokens[i]);
```

Maybe this would be a bit clearer given that it's not necessarily a pool.